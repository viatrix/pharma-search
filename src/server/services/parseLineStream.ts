import {AccessTypes, ApplTypes} from "../db/entity/products";
import {Transform, TransformCallback} from "stream";

export type FileProductType = {
  Ingredient: string,
  'DF;Route': string,
  Trade_Name: string,
  Applicant: string,
  Strength: string,
  Appl_Type: string,
  Appl_No: string,
  Product_No: string,
  TE_Code: string,
  Approval_Date: string,
  RLD: string,
  RS: string,
  Type: string,
  Applicant_Full_Name: string
}

export type ProductType = {
  ingredients: string[]
  dosageForm: string
  route: string
  tradeName: string
  applicant: string
  strength: string[]
  applType: ApplTypes
  applNo: number
  productNo: number
  teCode: string
  approvalDate: Date
  rld: boolean
  rs: boolean
  accessType: AccessTypes
  applicantFullName: string
}

export class ParseLineStream extends Transform {

  constructor() {
    super({objectMode: true});
  }

  public _transform(chunk: FileProductType, _encoding: string, callback: TransformCallback) {
    try {
      this.push(this.parseLine(chunk));
    }
    catch(error) {
      if (error.name !== 'InconsistentIngredientsStrength') {
        throw(error);
      }
    }
    callback();
  }

  private parseLine (line: FileProductType): ProductType {
    const FEDERAL_NOTE =
      ' **Federal Register determination that product was not discontinued or withdrawn for safety or efficacy reasons**';
    if (line.Strength.includes(FEDERAL_NOTE)) {
      line.Strength = line.Strength.slice(0, line.Strength.indexOf(FEDERAL_NOTE));
    }

    // Ingredients for each product are represented as a list separated by semicolon
    // Strength is specified for each ingredient
    // Lists of strength and ingredient have equal length
    // Strength and ingredients can have additional brackets, semicolons etc that require processing
    let ingredients: string[];
    if (line.Ingredient.includes(' (')) {
      const basicIngredients = line.Ingredient.slice(
          line.Ingredient.indexOf(' (') + 2,
          line.Ingredient.indexOf(')')
      );
      ingredients = basicIngredients.split(';');
    } else {
      ingredients = line.Ingredient.split('; ');
    }
    let strength = line.Strength.split(';');
    let balanced = true;
    for (let i = 0; i < strength.length; i++) {
      if (!this.bracketsBalanced(strength[i])) {
        balanced = false;
      }
    }
    let rawStrength = line.Strength;
    if (!balanced) {
      rawStrength = rawStrength.slice(
        rawStrength.indexOf(' (') + 2,
        rawStrength.indexOf(')')
      );
      strength = rawStrength.split(';');
    }
    if (strength.length !== ingredients.length) {
      const err = new Error(
        `Incorrect list of ingredients and strength for ${line.Trade_Name}`
      );
      err.name = 'InconsistentIngredientsStrength';
      throw err;
    }

    // Dosage form and route are specified as one field in csv separated by semicolon
    const dosageForm = line['DF;Route'].slice(0, line['DF;Route'].indexOf(';'));
    const route = line['DF;Route'].slice(line['DF;Route'].indexOf(';') + 1);

    // Application type and access type should be converted to enum
    let applType: ApplTypes;
    switch (line.Appl_Type) {
      case 'N':
        applType = ApplTypes.ORIGINAL;
        break;
      case 'A':
        applType = ApplTypes.GENERIC;
        break;
      default:
        throw new Error(`Invalid appl_type ${line.Appl_Type}`);
    }
    let accessType: AccessTypes;
    switch (line.Type) {
      case 'RX':
        accessType = AccessTypes.RX;
        break;
      case 'OTC':
        accessType = AccessTypes.OTC;
        break;
      case 'DISCN':
        accessType = AccessTypes.DISCN;
        break;
      default:
        throw new Error(`Invalid access_type ${line.Type}`);
    }

    const applNo = +line.Appl_No;
    const productNo = +line.Product_No;

    const approvalDate = new Date(line.Approval_Date);

    const rld = line.RLD === 'Yes';
    const rs = line.RS === 'Yes';

    return {
      ingredients,
      dosageForm,
      route,
      tradeName: line.Trade_Name,
      applicant: line.Applicant,
      strength,
      applType,
      applNo,
      productNo,
      teCode: line.TE_Code,
      approvalDate,
      rld,
      rs,
      accessType,
      applicantFullName: line.Applicant_Full_Name
    };
  }

  private bracketsBalanced (line: string): boolean {
    let balanced = 0;
    for (let i = 0; i < line.length; i++) {
      switch (line[i]) {
        case '(':
          balanced += 1;
          break;
        case ')':
          balanced -= 1;
      }
    }
    return balanced === 0;
  }
}