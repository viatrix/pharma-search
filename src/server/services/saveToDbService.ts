import {
  Connection,
  Repository
} from 'typeorm';
import 'reflect-metadata';
import { AccessTypes, ApplTypes, Product } from '../db/entity/products';
import { Applicant } from '../db/entity/applicants';
import { DosageForm } from '../db/entity/dosageForms';
import { Ingredient } from '../db/entity/ingredients';
import { ProductIngredient } from '../db/entity/productIngredients';

import { Route } from '../db/entity/routes';
import 'pg';
import { Writable } from 'stream';

export type ProductType = {
  ingredients: string[]
  dosageForm: string
  route: string
  tradeName: string
  applicant: string
  strength: string[]
  applType: ApplTypes
  applNo: number
  productNo: number
  teCode: string
  approvalDate: Date
  rld: boolean
  rs: boolean
  accessType: AccessTypes
  applicantFullName: string
}

export class SaveToDbService {
  private connection: Connection;

  private applicantsRepository: Repository<Applicant>;

  private dosageFormsRepository: Repository<DosageForm>;

  private ingredientsRepository: Repository<Ingredient>;

  private productIngredientsRepository: Repository<ProductIngredient>;

  private productsRepository: Repository<Product>;

  private routesRepository: Repository<Route>;

  constructor(connection: Connection) {
    this.connection = connection;
    this.applicantsRepository = this.connection.getRepository(Applicant);
    this.dosageFormsRepository = this.connection.getRepository(DosageForm);
    this.ingredientsRepository = this.connection.getRepository(Ingredient);
    this.productIngredientsRepository = this.connection.getRepository(
      ProductIngredient
    );
    this.productsRepository = this.connection.getRepository(Product);
    this.routesRepository = this.connection.getRepository(Route);
  }

  public async clearDb () {
    await this.productIngredientsRepository.delete({});
    await this.productsRepository.delete({});
    await this.applicantsRepository.delete({});
    await this.dosageFormsRepository.delete({});
    await this.ingredientsRepository.delete({});
    await this.routesRepository.delete({});
  }

  public saveToDbStream(): Writable {
    const outStream = new Writable({objectMode: true});
    outStream._write = (chunk, _encoding, callback: (error?: Error | null) => void) => {
      this.saveToDb(chunk).then(() => callback());
    };
    return outStream;
  }

  public async saveToDb (product: ProductType) {
    if (product.ingredients.length !== product.strength.length) {
      throw new Error(
        `Incorrect list of ingredients and strength for ${product.tradeName}`
      );
    }
    let applicantEntry = await this.applicantsRepository.findOne(
      product.applicant
    );
    if (!applicantEntry) {
      applicantEntry = await this.applicantsRepository.save({
        applicantShortName: product.applicant,
        applicantFullName: product.applicantFullName
      });
    }
    let dosageFormEntry = await this.dosageFormsRepository.findOne({
      where: { dosageForm: product.dosageForm }
    });
    if (!dosageFormEntry) {
      dosageFormEntry = await this.dosageFormsRepository.save({
        dosageForm: product.dosageForm
      });
    }
    let routeEntry = await this.routesRepository.findOne({
      where: { route: product.route }
    });
    if (!routeEntry) {
      routeEntry = await this.routesRepository.save({
        route: product.route
      });
    }
    const productEntry = await this.productsRepository.save({
      tradeName: product.tradeName,
      dosageForm: dosageFormEntry,
      route: routeEntry,
      applicant: applicantEntry,
      applType: product.applType,
      applNo: product.applNo,
      productNo: product.productNo,
      teCode: product.teCode,
      approvalDate: product.approvalDate,
      rld: product.rld,
      rs: product.rs,
      accessType: product.accessType
    });
    for (let i = 0; i < product.ingredients.length; i++) {
      let ingredientEntry = await this.ingredientsRepository.findOne({
        where: { ingredient: product.ingredients[i] }
      });
      if (!ingredientEntry) {
        ingredientEntry = await this.ingredientsRepository.save({
          ingredient: product.ingredients[i]
        });
      }
      await this.productIngredientsRepository.save({
        product: productEntry,
        ingredient: ingredientEntry,
        strength: product.strength[i]
      });
    }
  }
}
