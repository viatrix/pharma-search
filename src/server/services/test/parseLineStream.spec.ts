import { ParseLineStream } from '../parseLineStream';
import { PassThrough } from 'stream';
import { ProductType } from '../saveToDbService';

describe('ParseLineStream', () => {
  it('should parse one line of data', async () => {
    const parseLineStream = new ParseLineStream();
    const inputStream = new PassThrough({ objectMode: true });
    const outputStream = new PassThrough({ objectMode: true });
    const outObj: Array<ProductType> = [];
    outputStream._write = (chunk, _encoding, callback) => {
      outObj.push(chunk);
      callback();
    };
    const testData = {
      Ingredient: 'BUDESONIDE',
      'DF;Route': 'AEROSOL, FOAM;RECTAL',
      Trade_Name: 'UCERIS',
      Applicant: 'VALEANT PHARMS INTL',
      Strength: '2MG/ACTUATION',
      Appl_Type: 'N',
      Appl_No: '205613',
      Product_No: '001',
      TE_Code: '',
      Approval_Date: 'Oct 7, 2014',
      RLD: 'Yes',
      RS: 'Yes',
      Type: 'RX',
      Applicant_Full_Name: 'VALEANT PHARMACEUTICALS INTERNATIONAL'
    };
    await new Promise((resolve) => {
      outputStream.on('finish', () => {
        resolve();
      });
      inputStream.pipe(parseLineStream).pipe(outputStream);
      inputStream.write(testData);
      inputStream.end();
    });
    expect(outObj.length).toEqual(1);
    expect(outObj[0]).toMatchSnapshot();
  });

  it('should parse a line with complex ingredients', async () => {
    const parseLineStream = new ParseLineStream();
    const inputStream = new PassThrough({ objectMode: true });
    const outputStream = new PassThrough({ objectMode: true });
    const outObj: Array<ProductType> = [];
    outputStream._write = (chunk, _encoding, callback) => {
      outObj.push(chunk);
      callback();
    };
    const testData = {
      Ingredient: 'PANCRELIPASE (AMYLASE;LIPASE;PROTEASE)',
      'DF;Route': 'CAPSULE, DELAYED RELEASE;ORAL',
      Trade_Name: 'CREON',
      Applicant: 'ABBVIE',
      Strength: '60,000USP UNITS;12,000USP UNITS;38,000USP UNITS',
      Appl_Type: 'N',
      Appl_No: '020725',
      Product_No: '002',
      TE_Code: '',
      Approval_Date: 'Apr 30, 2009',
      RLD: 'Yes',
      RS: 'No',
      Type: 'RX',
      Applicant_Full_Name: 'ABBVIE INC'
    };
    await new Promise((resolve) => {
      outputStream.on('finish', () => {
        resolve();
      });
      inputStream.pipe(parseLineStream).pipe(outputStream);
      inputStream.write(testData);
      inputStream.end();
    });
    expect(outObj.length).toEqual(1);
    expect(outObj[0]).toMatchSnapshot();
  });

  it('should parse a line with braces in single ingredient', async () => {
    const parseLineStream = new ParseLineStream();
    const inputStream = new PassThrough({ objectMode: true });
    const outputStream = new PassThrough({ objectMode: true });
    const outObj: Array<ProductType> = [];
    outputStream._write = (chunk, _encoding, callback) => {
      outObj.push(chunk);
      callback();
    };
    const testData = {
      Ingredient: 'FERRIC HEXACYANOFERRATE(II)',
      'DF;Route': 'CAPSULE;ORAL',
      Trade_Name: 'RADIOGARDASE (PRUSSIAN BLUE)',
      Applicant: 'HEYL CHEMISCH',
      Strength: '500MG',
      Appl_Type: 'N',
      Appl_No: '021626',
      Product_No: '001',
      TE_Code: '',
      Approval_Date: 'Oct 2, 2003',
      RLD: 'Yes',
      RS: 'Yes',
      Type: 'RX',
      Applicant_Full_Name: 'HEYL CHEMISCH PHARMAZEUTISHE FABRIK'
    };
    await new Promise((resolve) => {
      outputStream.on('finish', () => {
        resolve();
      });
      inputStream.pipe(parseLineStream).pipe(outputStream);
      inputStream.write(testData);
      inputStream.end();
    });
    expect(outObj.length).toEqual(1);
    expect(outObj[0]).toMatchSnapshot();
  });

  it('should parse a line with complex strength', async () => {
    const parseLineStream = new ParseLineStream();
    const inputStream = new PassThrough({ objectMode: true });
    const outputStream = new PassThrough({ objectMode: true });
    const outObj: Array<ProductType> = [];
    outputStream._write = (chunk, _encoding, callback) => {
      outObj.push(chunk);
      callback();
    };
    const testData = {
      Ingredient: 'ARTICAINE HYDROCHLORIDE; EPINEPHRINE BITARTRATE',
      'DF;Route': 'INJECTABLE;INJECTION',
      Trade_Name: 'ARTICAINE HYDROCHLORIDE AND EPINEPHRINE BITARTRATE',
      Applicant: 'HOSPIRA',
      Strength: '4%;EQ 0.017MG BASE/1.7ML (4%;EQ 0.01MG BASE/ML)',
      Appl_Type: 'A',
      Appl_No: '079138',
      Product_No: '001',
      TE_Code: '',
      Approval_Date: 'Oct 2, 2003',
      RLD: 'No',
      RS: 'No',
      Type: 'DISCN',
      Applicant_Full_Name: 'HOSPIRA INC'
    };
    await new Promise((resolve) => {
      outputStream.on('finish', () => {
        resolve();
      });
      inputStream.pipe(parseLineStream).pipe(outputStream);
      inputStream.write(testData);
      inputStream.end();
    });
    expect(outObj.length).toEqual(1);
    expect(outObj[0]).toMatchSnapshot();
  });

  it('should parse a line with complex strength with brackets', async () => {
    const parseLineStream = new ParseLineStream();
    const inputStream = new PassThrough({ objectMode: true });
    const outputStream = new PassThrough({ objectMode: true });
    const outObj: Array<ProductType> = [];
    outputStream._write = (chunk, _encoding, callback) => {
      outObj.push(chunk);
      callback();
    };
    const testData = {
      Ingredient: 'OLIVE OIL; SOYBEAN OIL',
      'DF;Route': 'EMULSION;INTRAVENOUS',
      Trade_Name: 'CLINOLIPID 20%',
      Applicant: 'BAXTER HLTHCARE CORP',
      Strength: '16%(160GM/1000ML);4%  (40GM/1000ML)',
      Appl_Type: 'N',
      Appl_No: '204508',
      Product_No: '001',
      TE_Code: '',
      Approval_Date: 'Oct 3, 2013',
      RLD: 'Yes',
      RS: 'Yes',
      Type: 'RX',
      Applicant_Full_Name: 'HOSPIRA INC'
    };
    await new Promise((resolve) => {
      outputStream.on('finish', () => {
        resolve();
      });
      inputStream.pipe(parseLineStream).pipe(outputStream);
      inputStream.write(testData);
      inputStream.end();
    });
    expect(outObj.length).toEqual(1);
    expect(outObj[0]).toMatchSnapshot();
  });

  it('should remove federal note from strength', async () => {
    const parseLineStream = new ParseLineStream();
    const inputStream = new PassThrough({ objectMode: true });
    const outputStream = new PassThrough({ objectMode: true });
    const outObj: Array<ProductType> = [];
    outputStream._write = (chunk, _encoding, callback) => {
      outObj.push(chunk);
      callback();
    };
    const testData = {
      Ingredient: 'TRAZODONE HYDROCHLORIDEL',
      'DF;Route': 'TABLET, EXTENDED RELEASE;ORAL',
      Trade_Name: 'OLEPTRO',
      Applicant: 'ANGELINI PHARMA',
      Strength:
        '150MG **Federal Register determination that product was not discontinued or withdrawn for safety or efficacy reasons**',
      Appl_Type: 'N',
      Appl_No: '022411',
      Product_No: '001',
      TE_Code: '',
      Approval_Date: 'Feb 2, 2010',
      RLD: 'Yes',
      RS: 'No',
      Type: 'DISCN',
      Applicant_Full_Name: 'ANGELINI PHARMA INC'
    };
    await new Promise((resolve) => {
      outputStream.on('finish', () => {
        resolve();
      });
      inputStream.pipe(parseLineStream).pipe(outputStream);
      inputStream.write(testData);
      inputStream.end();
    });
    expect(outObj.length).toEqual(1);
    expect(outObj[0]).toMatchSnapshot();
  });

  it("should NOT parse file if appl_type is not 'N' or 'A'", async () => {
    const parseLineStream = new ParseLineStream();
    const inputStream = new PassThrough({ objectMode: true });
    const outputStream = new PassThrough({ objectMode: true });
    const outObj: Array<ProductType> = [];
    outputStream._write = (chunk, _encoding, callback) => {
      outObj.push(chunk);
      callback();
    };
    const testData = {
      Ingredient: 'BUDESONIDE',
      'DF;Route': 'AEROSOL, FOAM;RECTAL',
      Trade_Name: 'UCERIS',
      Applicant: 'VALEANT PHARMS INTL',
      Strength: '2MG/ACTUATION',
      Appl_Type: 'wrong_appl_type',
      Appl_No: '205613',
      Product_No: '001',
      TE_Code: '',
      Approval_Date: 'Oct 7, 2014',
      RLD: 'Yes',
      RS: 'Yes',
      Type: 'RX',
      Applicant_Full_Name: 'VALEANT PHARMACEUTICALS INTERNATIONAL'
    };
    try {
      await new Promise((resolve) => {
        outputStream.on('finish', () => {
          resolve();
        });
        inputStream.pipe(parseLineStream).pipe(outputStream);
        inputStream.write(testData);
        inputStream.end();
      });
    } catch (err) {
      expect(err.message).toEqual('Invalid appl_type wrong_appl_type');
    }
  });

  it('should NOT parse file if access type is not valid', async () => {
    const parseLineStream = new ParseLineStream();
    const inputStream = new PassThrough({ objectMode: true });
    const outputStream = new PassThrough({ objectMode: true });
    const outObj: Array<ProductType> = [];
    outputStream._write = (chunk, _encoding, callback) => {
      outObj.push(chunk);
      callback();
    };
    const testData = {
      Ingredient: 'BUDESONIDE',
      'DF;Route': 'AEROSOL, FOAM;RECTAL',
      Trade_Name: 'UCERIS',
      Applicant: 'VALEANT PHARMS INTL',
      Strength: '2MG/ACTUATION',
      Appl_Type: 'N',
      Appl_No: '205613',
      Product_No: '001',
      TE_Code: '',
      Approval_Date: 'Oct 7, 2014',
      RLD: 'Yes',
      RS: 'Yes',
      Type: 'wrong_access_type',
      Applicant_Full_Name: 'VALEANT PHARMACEUTICALS INTERNATIONAL'
    };
    try {
      await new Promise((resolve) => {
        outputStream.on('finish', () => {
          resolve();
        });
        inputStream.pipe(parseLineStream).pipe(outputStream);
        inputStream.write(testData);
        inputStream.end();
      });
    } catch (err) {
      expect(err.message).toEqual('Invalid access_type wrong_access_type');
    }
  });
});
