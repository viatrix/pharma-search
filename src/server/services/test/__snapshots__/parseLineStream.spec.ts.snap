// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParseLineStream should handle error 1`] = `
Object {
  "accessType": "RX",
  "applNo": 204508,
  "applType": "ORIGINAL",
  "applicant": "BAXTER HLTHCARE CORP",
  "applicantFullName": "HOSPIRA INC",
  "approvalDate": 2013-10-02T21:00:00.000Z,
  "dosageForm": "EMULSION",
  "ingredients": Array [
    "OLIVE OIL",
    "SOYBEAN OIL",
  ],
  "productNo": 1,
  "rld": true,
  "route": "INTRAVENOUS",
  "rs": true,
  "strength": Array [
    "16%(160GM/1000ML)",
    "4%  (40GM/1000ML)",
  ],
  "teCode": "",
  "tradeName": "CLINOLIPID 20%",
}
`;

exports[`ParseLineStream should parse a line with braces in single ingredient 1`] = `
Object {
  "accessType": "RX",
  "applNo": 21626,
  "applType": "ORIGINAL",
  "applicant": "HEYL CHEMISCH",
  "applicantFullName": "HEYL CHEMISCH PHARMAZEUTISHE FABRIK",
  "approvalDate": 2003-10-01T21:00:00.000Z,
  "dosageForm": "CAPSULE",
  "ingredients": Array [
    "FERRIC HEXACYANOFERRATE(II)",
  ],
  "productNo": 1,
  "rld": true,
  "route": "ORAL",
  "rs": true,
  "strength": Array [
    "500MG",
  ],
  "teCode": "",
  "tradeName": "RADIOGARDASE (PRUSSIAN BLUE)",
}
`;

exports[`ParseLineStream should parse a line with complex ingredients 1`] = `
Object {
  "accessType": "RX",
  "applNo": 20725,
  "applType": "ORIGINAL",
  "applicant": "ABBVIE",
  "applicantFullName": "ABBVIE INC",
  "approvalDate": 2009-04-29T21:00:00.000Z,
  "dosageForm": "CAPSULE, DELAYED RELEASE",
  "ingredients": Array [
    "AMYLASE",
    "LIPASE",
    "PROTEASE",
  ],
  "productNo": 2,
  "rld": true,
  "route": "ORAL",
  "rs": false,
  "strength": Array [
    "60,000USP UNITS",
    "12,000USP UNITS",
    "38,000USP UNITS",
  ],
  "teCode": "",
  "tradeName": "CREON",
}
`;

exports[`ParseLineStream should parse a line with complex strength 1`] = `
Object {
  "accessType": "DISCN",
  "applNo": 79138,
  "applType": "GENERIC",
  "applicant": "HOSPIRA",
  "applicantFullName": "HOSPIRA INC",
  "approvalDate": 2003-10-01T21:00:00.000Z,
  "dosageForm": "INJECTABLE",
  "ingredients": Array [
    "ARTICAINE HYDROCHLORIDE",
    "EPINEPHRINE BITARTRATE",
  ],
  "productNo": 1,
  "rld": false,
  "route": "INJECTION",
  "rs": false,
  "strength": Array [
    "4%",
    "EQ 0.01MG BASE/ML",
  ],
  "teCode": "",
  "tradeName": "ARTICAINE HYDROCHLORIDE AND EPINEPHRINE BITARTRATE",
}
`;

exports[`ParseLineStream should parse a line with complex strength with brackets 1`] = `
Object {
  "accessType": "RX",
  "applNo": 204508,
  "applType": "ORIGINAL",
  "applicant": "BAXTER HLTHCARE CORP",
  "applicantFullName": "HOSPIRA INC",
  "approvalDate": 2013-10-02T21:00:00.000Z,
  "dosageForm": "EMULSION",
  "ingredients": Array [
    "OLIVE OIL",
    "SOYBEAN OIL",
  ],
  "productNo": 1,
  "rld": true,
  "route": "INTRAVENOUS",
  "rs": true,
  "strength": Array [
    "16%(160GM/1000ML)",
    "4%  (40GM/1000ML)",
  ],
  "teCode": "",
  "tradeName": "CLINOLIPID 20%",
}
`;

exports[`ParseLineStream should parse one line of data 1`] = `
Object {
  "accessType": "RX",
  "applNo": 205613,
  "applType": "ORIGINAL",
  "applicant": "VALEANT PHARMS INTL",
  "applicantFullName": "VALEANT PHARMACEUTICALS INTERNATIONAL",
  "approvalDate": 2014-10-06T21:00:00.000Z,
  "dosageForm": "AEROSOL, FOAM",
  "ingredients": Array [
    "BUDESONIDE",
  ],
  "productNo": 1,
  "rld": true,
  "route": "RECTAL",
  "rs": true,
  "strength": Array [
    "2MG/ACTUATION",
  ],
  "teCode": "",
  "tradeName": "UCERIS",
}
`;

exports[`ParseLineStream should remove federal note from strength 1`] = `
Object {
  "accessType": "DISCN",
  "applNo": 22411,
  "applType": "ORIGINAL",
  "applicant": "ANGELINI PHARMA",
  "applicantFullName": "ANGELINI PHARMA INC",
  "approvalDate": 2010-02-01T22:00:00.000Z,
  "dosageForm": "TABLET, EXTENDED RELEASE",
  "ingredients": Array [
    "TRAZODONE HYDROCHLORIDEL",
  ],
  "productNo": 1,
  "rld": true,
  "route": "ORAL",
  "rs": false,
  "strength": Array [
    "150MG",
  ],
  "teCode": "",
  "tradeName": "OLEPTRO",
}
`;
